src/main/java/ch/njol/skript/sections/EffSecShoot.java

```diff
@@ -0,0 +1,298 @@
+				return false;
+			}
+		}
+		return true;
+	}
+
+	@Override
+	protected @Nullable TriggerItem walk(Event event) {
+		lastSpawned = null;
+		Number finalVelocity = velocity != null ? velocity.getSingle(event) : DEFAULT_SPEED;
+		Direction finalDirection = direction != null ? direction.getSingle(event) : Direction.IDENTITY;
+		if (finalVelocity == null || finalDirection == null)
+			return null;
+		EntityData<?>[] data = types.getArray(event);
+
+		for (Object shooter : shooters.getArray(event)) {
+			for (EntityData<?> entityData : data) {
+				Entity finalProjectile = null;
+				Vector vector;
+				if (shooter instanceof LivingEntity livingShooter) {
+					vector = finalDirection.getDirection(livingShooter.getLocation()).multiply(finalVelocity.doubleValue());
+					//noinspection rawtypes
+					Consumer afterSpawn = afterSpawn(event, entityData, livingShooter);
+					Class<? extends Entity> type = entityData.getType();
+					Location shooterLoc = livingShooter.getLocation();
+					shooterLoc.setY(shooterLoc.getY() + livingShooter.getEyeHeight() / 2);
+					boolean isProjectile = false, useWorldSpawn = false;
+					if (Fireball.class.isAssignableFrom(type)) {
+						shooterLoc = livingShooter.getEyeLocation().add(vector.clone().normalize().multiply(0.5));
+						isProjectile = true;
+						useWorldSpawn = true;
+					} else if (Projectile.class.isAssignableFrom(type)) {
+						isProjectile = true;
+						if (trigger != null && !RUNNING_PAPER) {
+							useWorldSpawn = true;
+						}
+					}
+
+					CaseUsage caseUsage = getCaseUsage(isProjectile, useWorldSpawn, trigger != null);
+
+					switch (caseUsage) {
+						case NOT_PROJECTILE_NO_TRIGGER -> {
+							finalProjectile = entityData.spawn(shooterLoc);
+						}
+						case NOT_PROJECTILE_TRIGGER -> {
+							//noinspection unchecked
+							entityData.spawn(shooterLoc, afterSpawn);
+						}
+						case PROJECTILE_NO_WORLD_NO_TRIGGER -> {
+							//noinspection unchecked
+							finalProjectile = livingShooter.launchProjectile((Class<? extends Projectile>) type);
+							set(finalProjectile, entityData);
+						}
+						case PROJECTILE_NO_WORLD_TRIGGER_BUKKIT -> {
+							try {
+								launchWithBukkitConsumer.invoke(livingShooter, type, vector, afterSpawnBukkit(event, entityData, livingShooter));
+							} catch (Exception ignored) {};
+						}
+						case PROJECTILE_NO_WORLD_TRIGGER -> {
+							//noinspection unchecked
+							livingShooter.launchProjectile((Class<? extends Projectile>) type, vector, afterSpawn);
+						}
+						case PROJECTILE_WORLD_NO_TRIGGER -> {
+							Projectile projectile = (Projectile) livingShooter.getWorld().spawn(shooterLoc, type);
+							projectile.setShooter(livingShooter);
+							finalProjectile = projectile;
+						}
+						case PROJECTILE_WORLD_TRIGGER -> {
+							//noinspection unchecked
+							livingShooter.getWorld().spawn(shooterLoc, type, afterSpawn);
+						}
+						default -> throw new IllegalStateException("Unexpected value: " + caseUsage);
```

Comment:
This case usage stuff seems kind of messy and this method is becoming a bit too long and complex. The cyclomatic complexity is ridiculously high (somewhere over 50 by my headcount) and I think it needs to be broken up.

You've got a few options (e.g. move part of the method from the loop into its own method) but personally I would recommend that you make the CaseUsage a functional interface (accepting the required variables) and then you can move each application into its own space and eliminate the switch entirely.